python/lsst/daf/butler/script/queryDatasets.py

```diff
@@ -225,30 +239,59 @@ def getDatasets(self) -> Iterator[DatasetRef]:
+        dataset_types = set(
+            self.butler.collections._filter_dataset_types(dataset_types, query_collections_info)
+        )
 
-            if (n_filtered := len(dataset_types)) != n_dataset_types:
-                _LOG.info("Filtered %d dataset types down to %d", n_dataset_types, n_filtered)
-            elif n_dataset_types == 0:
-                _LOG.info("The given dataset type, %s, is not known to this butler.", datasetTypes)
-            else:
-                _LOG.info(
-                    "Processing %d dataset type%s", n_dataset_types, "" if n_dataset_types == 1 else "s"
-                )
-
-            # Accumulate over dataset types.
-            for dt in sorted(dataset_types):
-                results = query.datasets(dt, collections=query_collections, find_first=self._find_first)
-                if self._where:
-                    results = results.where(self._where)
-                yield from results.with_dimension_records()
+        if (n_filtered := len(dataset_types)) != n_dataset_types:
+            _LOG.info("Filtered %d dataset types down to %d", n_dataset_types, n_filtered)
+        elif n_dataset_types == 0:
+            _LOG.info("The given dataset type, %s, is not known to this butler.", datasetTypes)
+        else:
+            _LOG.info("Processing %d dataset type%s", n_dataset_types, "" if n_dataset_types == 1 else "s")
+
+        # Accumulate over dataset types.
+        limit = self._limit
+        warn_limit = False
+        unlimited = True if limit == 0 else False
+        if limit < 0:
+            warn_limit = True
+            limit = abs(limit) + 1  # +1 to tell us we hit the limit.
+        for dt in sorted(dataset_types):
+            kwargs: dict[str, Any] = {}
+            if self._where:
+                kwargs["where"] = self._where
+            if not unlimited:
+                kwargs["limit"] = limit
+            _LOG.debug("Querying dataset type %s with %s", dt, kwargs)
+            results = self.butler.query_datasets(
+                dt,
+                collections=query_collections,
+                find_first=self._find_first,
+                with_dimension_records=True,
+                order_by=self._order_by,
+                **kwargs,
+            )
+            if not unlimited:
+                limit -= len(results)
+                if warn_limit and limit == 0 and results:
+                    # We asked for one too many so must remove that from
+                    # the list.
+                    results.pop(-1)
+            _LOG.debug("Got %d results for dataset type %s", len(results), dt)
+            yield from results
+
+            if not unlimited and limit == 0:
+                if warn_limit:
+                    _LOG.warning(
+                        "Requested limit of %d hit for number of datasets returned. "
+                        "Use --limit to increase this limit.",
+                        self._limit,
+                    )
```

Comment:
I think we should move this warn-if-limit-hit logic into the Python API, even if we have to do it here too (at least until we implement multiple dataset type queries as a new Python interface).  That could be another ticket, of course.

Reply:
Ok. I can remove this warning completely here if the python API is going to handle it.

Reply:
I added the warning into `Butler.query_datasets` but had to leave the warning here because the context is different. If you ask for 1000 results spread over 20 dataset types you want to know the limit is 1000 but the internal query only thinks the limit was 100 or whatever.